Generics.md

```diff
@@ -1,164 +1,386 @@
-
-* generic values supplied with imports are single world tokens (typically wgsl type names or function names), 
-  or generic variables declared on that function.
+# Generic Declarations and Generic Modules
 
-## Examples
+I would like to limit as much as possible the amount of additional syntax and semantics in our extensions.
+After all, we are writing an extension, not a different language!
+It should look and feel just like the original stuff and extensions should be straight forward, self-explanatory even.
+
+## The proposal
+
+It builds on 3 observations
+
+* WGSL *already* has support for parametric types and functions, they are called [type-generators](https://www.w3.org/TR/WGSL/#type-generator) and [function overloads](https://www.w3.org/TR/WGSL/#builtin-functions). It even has some syntax for them.
+  The issue is, they are not allowed to be declared in user code. This proposal changes this.
+* A *module* is a collection of functions, structs and associated types that share a semantic. Turns out, a WGSL file is just that. This proposal introduces *implicit modules* as files.
+* global value and alias declarations are immutable and must be initialized. What if we could override them? if so, this is a simple mechanism for module specialization.
+  What if we also could override struct and functions declarations? In fact, all declarations?
+
+It proposes two extensions: [**Generic Declarations**](#generic-declarations) and [**Generic Modules**](#generic-modules) (+ optional [**Function-as-Type**](#function-as-type) [**Module-as-Type**](#module-as-type)).
+They are orthogonal but they share the concepts of [*Type Hierarchy*](formalism-type-hierarchy) and [*Generic Type*](#formalism-generic-type) formalized below.
+We could support only one of the two and achieve the same expressivity AFAIK. But not the same ergonomics.
+
+## Formalism: Type Hierarchy
+
+Let's introduce the concept of type hierarchy. From the [conversion ranks](https://www.w3.org/TR/WGSL/#conversion-rank) we can deduce the type hierarchy for built-in types already:
+
+* `u32` and `i32` are subtypes of `AbstractInt`, because `AbstractInt` can be converted to `u32` and `f32`
+* `f32` and `f16` are subtypes of `AbstractFloat`, because `AbstractInt` can be converted to `u32` and `f32`
+* `AbstractInt` is a subtype of `AbstractFloat`, because `AbstractInt` can be converted to] `AbstractFloat`
+* `VecN<S>` is a subtype of `VecN<T>` if `S` is a subtype of `T`
+* `MatCxR<S>` is a subtype of `MatCxR<T>` if `S` is a subtype of `T`
+* `Array<S,N>` is a subtype of `Array<T,N>` if `S` is a subtype of `T`
+
+And we can add to that list:
+* `atomic<S>` is a subtype of `atomic<T>` if `S` is a subtype of `T`.
+* `ptr<S>` is a subtype of `ptr<T>` if `S` is a subtype of `T`.
+
+We can extend subtyping for user-defined types:
+* structure type `S1` is a subtype of structure type `S2` if, for all members `m` in `S2`: `S1.m` exists and `S1.m` is a subtype of `S2.m` (member covariance).
```

Comment:
This is certainly an interesting property, but also feels a bit dangerous because of how  memory layout would differ, potentially impacting stuff like arrays, vecs, etc